syftlib/src/main/java/org/openmined/syft/Syft.kt

```diff
@@ -1,85 +1,139 @@
+        val job = SyftJob(this, model, version)
+        val jobId = SyftJob.JobID(model, version)
+        workerJobs[jobId] = job
+        compositeDisposable.add(job.getStatusProcessor()
+                .compose(schedulers.applyFlowableSchedulers())
+                .subscribe(
+                    {},
+                    { workerJobs.remove(jobId) },
+                    { workerJobs.remove(jobId) }
+                )
+        )
 
-                        is NetworkMessage.SocketClosed -> Log.d(
-                            TAG,
-                            "Socket was closed successfully"
-                        )
-                        is NetworkMessage.SocketError -> Log.e(TAG, "socket error", it.throwable)
-                        is NetworkMessage.MessageReceived -> handleResponse(
-                            CommunicationDataFactory.deserializeSocket(
-                                it.message
-                            )
-                        )
-                        is NetworkMessage.MessageSent -> println("Message sent successfully")
+        return job
+    }
+
+    fun requestCycle(job: SyftJob) {
+        if (this::workerId.isInitialized)
+            socketClient.getCycle(
+                CycleRequest(
+                    workerId,
+                    job.modelName,
+                    job.version,
+                    getPing(),
+                    getDownloadSpeed(),
+                    getUploadSpeed()
+                )
+            ).compose(schedulers.applySingleSchedulers())
+        else {
+            compositeDisposable.add(socketClient.authenticate()
+                    .compose(schedulers.applySingleSchedulers())
+                    .subscribe { t: AuthenticationSuccess ->
+                        if (!this::workerId.isInitialized)
+                            setSyftWorkerId(workerId)
+                        requestCycle(job)
                     }
-                }
-                .subscribeOn(schedulers.computeThreadScheduler)
-                .observeOn(schedulers.calleeThreadScheduler)
-                .subscribe()
-    )
+            )
+        }
+    }
 
-    fun newJob(modelName: String, version: String): SyftJob {
-        val job = SyftJob(modelName, version)
-        signallingClient.send(
-            REQUESTS.CYCLE_REQUEST,
-            CommunicationDataFactory.requestCycle(workerId, job, "", "", "")
-        )
-        workerJobs.add(job)
-        return job
+    @Synchronized
+    private fun setSyftWorkerId(workerId: String) {
+        if (!this::workerId.isInitialized)
+            this.workerId = workerId
     }
 
+    private fun getPing() = ""
+    private fun getDownloadSpeed() = ""
+    private fun getUploadSpeed() = ""
+
     private fun handleResponse(response: SocketResponse) {
         when (response.data) {
-            is AuthenticationSuccess ->
+            is AuthenticationSuccess -> {
                 this.workerId = response.data.workerId
```

Comment:
Could use here `setSyftWorkerId(response.data.workerId)` or it doesn't matter whether the `workerId` had been initialised before in this case?

Reply:
There are bigger problems with worker id setting which I am currently resolving. We would need to ensure that before changing the worker id all the jobs using the earlier one are finished and reported. setWorkerId isnt finalized yet